src/create_torrent.cpp

```diff
@@ -650,42 +746,72 @@ namespace {
 					}
 				}
 			}
-		}
 
-		info["piece length"] = m_files.piece_length();
-		if (m_merkle_torrent)
-		{
-			int const num_leafs = merkle_num_leafs(m_files.num_pieces());
-			int const num_nodes = merkle_num_nodes(num_leafs);
-			int const first_leaf = num_nodes - num_leafs;
-			m_merkle_tree.resize(num_nodes);
-			auto const num_pieces = int(m_piece_hash.size());
-			for (int i = 0; i < num_pieces; ++i)
-				m_merkle_tree[first_leaf + i] = m_piece_hash[piece_index_t(i)];
-			for (int i = num_pieces; i < num_leafs; ++i)
-				m_merkle_tree[first_leaf + i].clear();
-
-			// now that we have initialized all leaves, build
-			// each level bottom-up
-			int level_start = first_leaf;
-			int level_size = num_leafs;
-			while (level_start > 0)
+			if (!m_v1_only && !info.find_key("file tree"))
 			{
-				int parent = merkle_get_parent(level_start);
-				for (int i = level_start; i < level_start + level_size; i += 2, ++parent)
+				auto& tree = info["file tree"];
+
+				for (file_index_t i(0); i != m_files.end_file(); ++i)
 				{
-					hasher h;
-					h.update(m_merkle_tree[i]);
-					h.update(m_merkle_tree[i + 1]);
-					m_merkle_tree[parent] = h.final();
+					if (files().file_flags(i) & file_storage::flag_pad_file) continue;
+
+					entry* file_e_ptr = &tree;
+
+					{
+						std::string const file_path = m_files.file_path(i);
+						auto const split = lsplit_path(file_path);
+						TORRENT_ASSERT(split.first == m_files.name());
+
+						for (auto e = lsplit_path(split.second);
+							!e.first.empty();
+							e = lsplit_path(e.second))
+						{
+							file_e_ptr = &(*file_e_ptr)[e.first];
+							if (file_e_ptr->dict().find({}) != file_e_ptr->dict().end())
+							{
+								// path conflict
+								// there is already a file with this name
+								// refuse to generate a torrent with such a conflict
+								// TODO: 3 report error
+								return entry();
+							}
+						}
+					}
+
+					if (!file_e_ptr->dict().empty())
+					{
+						// path conflict
+						// there is already a directory with this name
+						// refuse to generate a torrent with such a conflict
+						// TODO: 3 report error
```

Comment:
I think an exception is reasonable here, and the case above. What do you think? I can prepare a patch

Reply:
sgtm